strawberryfields/apps/train/cost.py

```diff
@@ -118,3 +120,194 @@ def evaluate(self, params: np.ndarray) -> float:
+
+        Returns:
+            float: the value of the stochastic cost function
+        """
+        samples = self.vgbs.get_A_init_samples(n_samples)
+        return np.mean([self.h_reparametrized(s, params) for s in samples])
+
+    def h_reparametrized(self, sample: np.ndarray, params: np.ndarray) -> float:
+        r"""Include trainable parameters in the :math:`h(\bar{n})` function to allow sampling
+        from the initial adjacency matrix.
+
+        The reparametrized function can be written in terms :math:`h(\bar{n})` as:
+
+        .. math::
+
+            h(\bar{n}, \theta) = h(\bar{n}) \sqrt{\frac{\det (\mathbb{I} - A(\theta)^{2})}
+            {\det (\mathbb{I} - A^{2})}} \prod_{k=1}^{m}w_{k}^{n_{k}},
+
+        where :math:`w_{k}` is the :math:`\theta`-dependent weight on the :math:`k`-th mode in
+        the :class:`~.train.VGBS` system and :math:`n_{k}` is the number of photons in mode :math:`k`.
+
+        **Example usage:**
+
+        >>> sample = [1, 1, 0, 0]
+        >>> cost.h_reparametrized(sample, params)
+        -1.6688383062813434
+
+        Args:
+            sample (array): the sample
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            float: the cost function with respect to a given sample and set of trainable parameters
+        """
+        h = self.h(sample)
+        A = self.vgbs.A(params)
+        w = self.vgbs.embedding(params)
+        Id = np.eye(2 * self.vgbs.n_modes)
+
+        dets_numerator = np.linalg.det(Id - _Omat(A))
+        dets_denominator = np.linalg.det(Id - _Omat(self.vgbs.A_init))
+        dets = np.sqrt(dets_numerator / dets_denominator)
+
+        prod = np.prod(np.power(w, sample))
+
+        return h * dets * prod
+
+    def _gradient_one_sample(self, sample: np.ndarray, params: np.ndarray) -> np.ndarray:
+        """Evaluates the gradient equation on a single sample.
+
+        Args:
+            sample (array): the sample
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the one-shot gradient
+        """
+        w = self.vgbs.embedding(params)
+        jac = self.vgbs.embedding.jacobian(params)
+
+        h = self.h_reparametrized(sample, params)
+
+        if self.vgbs.threshold:
+            diff = sample - self.vgbs.mean_clicks_by_mode(params)
+        else:
+            diff = sample - self.vgbs.mean_photons_by_mode(params)
+
+        return h * (diff / w) @ jac
+
+    def gradient(self, params: np.ndarray, n_samples: int) -> np.ndarray:
+        r"""Evaluates the gradient of the cost function.
+
+        As shown in `this paper <https://arxiv.org/abs/2004.04770>`__, the gradient can be
+        evaluated by finding an average over samples generated from the input adjacency matrix to
+        the VGBS system:
+
+        .. math::
+
+            \partial_{\theta} C (\theta) = \sum_{\bar{n}} h(\bar{n}, \theta) P(\bar{n})
```

Comment:
Perhaps I missed something, I just came across that this would be an approximate
```suggestion
            \partial_{\theta} C (\theta) \approx \sum_{\bar{n}} h(\bar{n}, \theta) P(\bar{n})
```